gitlab.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		return nil, fmt.Errorf("GitLabMergeRequest needs 'git' command: %v", err)
+	}
+	return &GitLabMergeRequest{
+		cli:      cli,
+		owner:    owner,
+		repo:     repo,
+		pr:       pr,
+		sha:      sha,
+		projects: owner + "/" + repo,
+		wd:       workDir,
+	}, nil
+}
+
+// Post accepts a comment and holds it. Flush method actually posts comments to
+// GitLab in parallel.
+func (g *GitLabMergeRequest) Post(_ context.Context, c *Comment) error {
+	c.Path = filepath.Join(g.wd, c.Path)
+	g.muComments.Lock()
+	defer g.muComments.Unlock()
+	g.postComments = append(g.postComments, c)
+	return nil
+}
+
+// Flush posts comments which has not been posted yet.
+func (g *GitLabMergeRequest) Flush(ctx context.Context) error {
+	defer g.muComments.Unlock()
+	g.muComments.Lock()
+
+	if err := g.setPostedComment(ctx); err != nil {
+		return err
+	}
+
+	return g.postCommentsForEach(ctx)
+}
+
+func (g *GitLabMergeRequest) postCommentsForEach(ctx context.Context) error {
+	var eg errgroup.Group
+	for _, c := range g.postComments {
+		comment := c
+		if g.postedcs.IsPosted(comment, comment.Lnum) {
+			continue
+		}
+		eg.Go(func() error {
+			commitID, err := g.getLastCommitsID(comment.Path, comment.Lnum)
+			if err != nil {
+				return err
+			}
+			body := commentBody(comment)
+			ltype := "new"
+			prcomment := &gitlab.PostCommitCommentOptions{
+				Note:     &body,
+				Path:     &comment.Path,
+				Line:     &comment.Lnum,
+				LineType: &ltype,
+			}
+			_, _, err = g.cli.Commits.PostCommitComment(g.projects, commitID, prcomment, nil)
+			return err
+		})
+	}
+	return eg.Wait()
+}
+
+func (g *GitLabMergeRequest) getLastCommitsID(path string, line int) (string, error) {
+	lineFormat := fmt.Sprintf("%d,%d", line, line)
+	s, err := exec.Command("git", "blame", "-l", "-L", lineFormat, path).Output()
+	if err != nil {
+		return "", fmt.Errorf("failed to get commitID: %v", err)
+	}
+	commitID := strings.Split(string(s), " ")[0]
+	return commitID, nil
+}
+
+func (g *GitLabMergeRequest) setPostedComment(ctx context.Context) error {
+	g.postedcs = make(postedcomments)
+	cs, err := g.comment(ctx)
+	if err != nil {
+		return err
+	}
+	for _, c := range cs {
+		if &c.Line == nil || &c.Path == nil || &c.Note == nil {
+			// skip resolved comments. Or comments which do not have "path" nor
+			// "body".
+			continue
+		}
+		path := c.Path
+		pos := c.Line
+		body := c.Note
+		if _, ok := g.postedcs[path]; !ok {
+			g.postedcs[path] = make(map[int][]string)
+		}
+		if _, ok := g.postedcs[path][pos]; !ok {
+			g.postedcs[path][pos] = make([]string, 0)
+		}
+		g.postedcs[path][pos] = append(g.postedcs[path][pos], body)
+	}
+	return nil
+}
+
+// Diff returns a diff of MergeRequest. It runs `git diff` locally instead of
+// diff_url of GitLab Pull Request because diff of diff_url is not suited for
+// comment API in a sense that diff of diff_url is equivalent to
+// `git diff --no-renames`, we want diff which is equivalent to
+// `git diff --find-renames`.
```

Comment:
[just a comment] hmm.. that's sad... I also experienced same problem with GitHub API :cry:

I reported this problem to GitHub and it's now fixed. Do you report this problem to GitLab or do you know are there any filed bug?